src/storage/cache/rpc/client.rs

```diff
@@ -0,0 +1,458 @@
+};
+
+/// TODO: combine RpcClientConnectionInner and RpcClientConnection
+struct RpcClientConnectionInner<P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    /// The TCP stream for the connection.
+    stream: UnsafeCell<net::TcpStream>,
+    /// Options for the timeout of the connection
+    timeout_options: TimeoutOptions,
+    /// Stream auto increment sequence number, used to mark the request and response
+    seq: AtomicU64,
+    /// Received keep alive seq number, will mark the valid keep alive response.
+    /// If the keep alive response is not received in 100 times, the connection will be closed,
+    /// If we receive other response, we will update the received_keepalive_seq too, just treat it as a keep alive response.
+    received_keepalive_seq: AtomicU64,
+    /// Send packet task
+    packets_keeper: UnsafeCell<PacketsKeeper<P>>,
+    /// Response buffer, try to reuse same buffer to reduce memory allocation
+    /// In case of the response is too large, we need to consider the buffer size
+    /// Init size is 4MB
+    /// Besides, we want to reduce memory copy, and read/write the response to the same data buffer
+    resp_buf: UnsafeCell<BytesMut>,
+}
+
+impl<P> RpcClientConnectionInner<P>
+where
+    P: Packet + Clone + Send + Sync + 'static
+{
+    pub fn new(stream: net::TcpStream, timeout_options: TimeoutOptions) -> Self {
+        Self {
+            stream: UnsafeCell::new(stream),
+            timeout_options: timeout_options.clone(),
+            seq: AtomicU64::new(0),
+            received_keepalive_seq: AtomicU64::new(0),
+            packets_keeper: UnsafeCell::new(PacketsKeeper::new(timeout_options.clone().idle_timeout.as_secs())),
+            resp_buf: UnsafeCell::new(BytesMut::with_capacity(8 * 1024 * 1024)),
+        }
+    }
+
+    /// Recv request header from the stream
+    pub async fn recv_header(&self) -> Result<RespHeader, RpcError<String>> {
+        // Try to read to buffer
+        match self.recv_len(REQ_HEADER_SIZE).await {
+            Ok(_) => {}
+            Err(err) => {
+                debug!("Failed to receive request header: {:?}", err);
+                return Err(err);
+            }
+        }
+
+        let buffer: &mut BytesMut = unsafe {
+            transmute(self.resp_buf.get())
+        };
+        let req_header = RespHeader::decode(&buffer)?;
+        debug!("Received response header: {:?}", req_header);
+
+        Ok(req_header)
+    }
+
+    /// Receive request body from the server.
+    pub async fn recv_len(&self, len: u64) -> Result<(), RpcError<String>> {
+        let mut req_buffer: &mut BytesMut = unsafe {
+            transmute(self.resp_buf.get())
+        };
+        req_buffer.resize(len as usize, 0);
+        let reader = self.get_stream_mut();
+        match read_exact_timeout!(reader, &mut req_buffer, self.timeout_options.read_timeout).await
+        {
+            Ok(size) => {
+                debug!("Received response body len: {:?}", size);
```

Comment:
All log related to client should include information about "self"  (eg: destination), there're multiple rpc client instances exist in one program.

Reply:
Ok, I have add a `Debug` trait and add `self` debug.